rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,526 @@
+
+This IPC API will be exposed from the clustering service.
+
+```ts
+export interface ClusteringServiceSetup {
+  // [...]
+  broadcast: (type: string, payload?: ClusterMessagePayload, options?: BroadcastOptions) => void;
+  addMessageHandler: (type: string, handler: MessageHandler) => MessageHandlerUnsubscribeFn;
+}
+```
+
+Notes:
+-  to reduce clustered and non-clustered mode divergence, in non-clustered mode, these APIs would just be no-ops. 
+   It will avoid to force (most) code to check in which mode Kibana is running before calling them.
+-  We could eventually use an Observable pattern instead of a handler pattern to subscribe to messages.
+
+## Executing code on a single worker
+
+In some scenarios, we would like to have parts of the code executed only from a single process. 
+
+The SO migration would be a good example: 
+we don't need to have each worker try to perform the migration, and we'd prefer to have one performing/trying the migration, 
+and the others wait for it. Due to the architecture, we can't have the coordinator perform such single-process jobs, 
+as it doesn't actually run a Kibana server.
+
+There are various ways to address such use-cases. What seems to be the best compromise right now would be the concept of 
+'main worker'. The coordinator would arbitrary elect a worker as the 'main' one at startup. The clustering service would then expose 
+an API to let workers identify themselves as main or not.
+
+```ts
+export interface ClusteringServiceSetup {
+  // [...]
+  isMainWorker: () => boolean;
+}
+```
+
+Notes:
+- in non-clustered mode, `isMainWorker` would always return true, to reduce the divergence between clustered and 
+  non-clustered modes.
+
+### Example: SO Migration
+
+To take the example of SO migration, the `KibanaMigrator.runMigrations` implementation could change to
+(naive implementation, the function is supposed to return a promise here, did not include that for simplicity):
+
+```ts
+runMigration() {
+   if(clustering.isMainWorker()) {
+     this.runMigrationsInternal().then((result) => {
+        applyMigrationState(result);
+        clustering.broadcast('migration-complete', { payload: result }, { persist: true }); // persist: true will send message even if subscriber subscribe after the message was actually sent
+      })
+   }
+   else {
+     const unsubscribe = clustering.addMessageHandler('migration-complete', ({ payload: result }) => {
+       applyMigrationState(result);
+       unsubscribe();
+     });
+   }
+}
+```
+
+Notes:
+  - to be sure that we do not encounter a race condition with the event subscribing / sending (workers subscribing after 
+    the main worker actually send the `migration-complete` event and then waiting indefinitely), we are using the `persist` option of the `broadcast` API.
+    I'm honestly not sure this is the best pattern, but I couldn't come with anything better (maybe shared state instead?). 
+    This is probably something else we should discuss.
```

Comment:
I prefer the `persist` option on messages over using shared state. Main reason being that we can easily log it and trace the order of events & execution to debug issues that arise.